src/r1cs_to_qap.rs

```diff
@@ -194,16 +196,27 @@ impl R1CStoQAP for LibsnarkReduction {
             });
 
         domain.ifft_in_place(&mut c);
-        domain.coset_fft_in_place(&mut c);
+        coset_domain.fft_in_place(&mut c);
 
         cfg_iter_mut!(ab)
             .zip(c)
             .for_each(|(ab_i, c_i)| *ab_i -= &c_i);
 
-        domain.divide_by_vanishing_poly_on_coset_in_place(&mut ab);
-        domain.coset_ifft_in_place(&mut ab);
+        let vanishing_polynomial = domain.vanishing_polynomial();
+
+        let inverted_polynomial_points: Vec<F> = coset_domain
```

Comment:
all these points are equal to domain.evaluate_vanishing_polynomial(F::GENERATOR).inverse()

Reply:
I tried to use evaluate_vanishing_ploynomial before, but with then my groth16 test failed, with the current implementation, it passes.

What I don't understand: `domain.evaluate_vanishing_polynomial(F::GENERATOR).inverse()` is returning a value in `F` and not in `Vec<F>`. Shouldn't we get inverses of an evaluation of the polynomial in all points `\omega_i` of the FFT basis of the domain, translated by `F::GENERATOR`?

Reply:
Let's to keep it simple assume a radix-2 domain (multiplicative subgroup) `H={w^i}, i=1,...,N`. It's vanishing polynomial is `z(X) = X^N - 1`. A coset of the domain is `gH = {g w^i}`, for some `g`. We want to divide by the vanishing polynomial of the domain in the evaluation form (aka in the Lagrange basis). By definition z is 0 in points of H, so we will use evaluations on a coset, they are: `z(g w^i) = (g w^i)^N - 1 = g^N - 1`.

Reply:
Ah, because the Fourier basis consists of roots of unity? But that doesn't always have to be the case?

Reply:
It's definitely the case for all domains implemented by arkworks. It may also hold for most practical domains. For a subgroup, the vanishing polynomial is probably still of the form `zH(X) = X^N - 1`, and for a coset of a subgroup `zgH(X) = prod (X - gh) = prod g(X/g - h) = g^N prod (X/g - h) = g^N zH(X/g) = X^N - g^N`.